### PR TITLE
Scripts/Battlegrounds: Implement Silvershard mines

### PR DESCRIPTION
**Pull request by Golrag**
**Changes proposed:**

-  Implement Silvershard Mines
-  Criteria Type KillPlayer was not being called in battleground. This is required for an achievement so I enabled it

**Issues addressed:**

Closes: None

**Tests performed:**

- Builds locally
- tested in game by playing the battleground

**Known issues and TODO list:** (add/remove lines as needed)

- [x] [Capturing a mine cart is too slow due to missing feature in control zone handling](https://github.com/TrinityCore/TrinityCore/pull/30152)
- [x] https://github.com/TrinityCore/TrinityCore/pull/30083
- [x] Battleground should end when team gets max score (currently only ends when mine cart is captured and score is reached)

### CODE REVIEW

File: src/server/game/Entities/GameObject/GameObject.cpp [hunk 1]
```diff
@@ -4477,6 +4477,18 @@ void GameObject::HandleCustomTypeCommand(GameObjectTypeBase::CustomCommand const
         command.Execute(*m_goTypeImpl);
 }
 
+TeamId GameObject::GetControllingTeam() const
+{
+    if (GetGoType() != GAMEOBJECT_TYPE_CONTROL_ZONE)
+        return TEAM_NEUTRAL;;
```
Comment (jackpoz) [inline src/server/game/Entities/GameObject/GameObject.cpp:4483]:
```suggestion
        return TEAM_NEUTRAL;
```

File: src/server/scripts/Battlegrounds/SilvershardMines/battleground_silvershard_mines.cpp [hunk 2]
```diff
@@ -0,0 +1,1011 @@
...
+    GuidSet _theLongRidersPlayers;
+    time_t _theLongRidersSpawnTime;
+};
+
+// 115904 - Cart Cap
+class spell_bg_silvershard_mines_cart_cap final : public SpellScript
+{
+    bool Validate(SpellInfo const* spellInfo) override
+    {
+        return ValidateSpellEffect({ { spellInfo->Id, EFFECT_0 } });
```
Comment (Nyr97) [inline src/server/scripts/Battlegrounds/SilvershardMines/battleground_silvershard_mines.cpp:848]: You don't need to validate an effect you're currently hooking, it's done internally.

File: src/server/scripts/Battlegrounds/SilvershardMines/battleground_silvershard_mines.cpp [hunk 3]
```diff
@@ -0,0 +1,1011 @@
...
+        OnEffectHitTarget += SpellEffectFn(spell_bg_silvershard_mines_cart_cap::HandleScript, EFFECT_0, SPELL_EFFECT_SCRIPT_EFFECT);
+    }
+};
+
+// 124491 - capturing
+class spell_bg_silvershard_mines_capturing final : public SpellScript
+{
+    bool Validate(SpellInfo const* spellInfo) override
+    {
+        return ValidateSpellEffect({ { spellInfo->Id, EFFECT_0 } })
```
Comment (Nyr97) [inline src/server/scripts/Battlegrounds/SilvershardMines/battleground_silvershard_mines.cpp:877]: Same thing here, we don't need to validate an effect we're hooking.

File: src/server/scripts/Battlegrounds/SilvershardMines/battleground_silvershard_mines.cpp [hunk 4]
```diff
@@ -0,0 +1,1011 @@
...
+        OnEffectHitTarget += SpellEffectFn(spell_bg_silvershard_mines_capturing::HandleScript, EFFECT_0, SPELL_EFFECT_SCRIPT_EFFECT);
+    }
+};
+
+// 97372 - Capturing
+class spell_bg_silvershard_mines_capturing_switch_track final : public SpellScript
+{
+    bool Validate(SpellInfo const* spellInfo) override
+    {
+        return ValidateSpellEffect({ { spellInfo->Id, EFFECT_0 } })
```
Comment (Nyr97) [inline src/server/scripts/Battlegrounds/SilvershardMines/battleground_silvershard_mines.cpp:898]: Same thing here, we don't need to validate an effect we're hooking.

File: src/server/scripts/Battlegrounds/SilvershardMines/battleground_silvershard_mines.cpp [hunk 5]
```diff
@@ -0,0 +1,1011 @@
...
+                return SilvershardMines::Spells::MineCartCourierNorthWest;
+            default:
+                return 0;
+        }
+    }
+
+    uint32 GetTheLongRiderAchievementSpell(uint32 pathId) const
+    {
+        TeamId controllingTeam = TEAM_NEUTRAL;
+        if (GameObject const* cz = GetControlZone())
```
Comment (Nyr97) [inline src/server/scripts/Battlegrounds/SilvershardMines/battleground_silvershard_mines.cpp:791]: cz might not be as obvious as bg for a pointer name, just a small note.

**Pull request merged**